### ISSUE TITLE
Memory Limit Exceeded Support

### ISSUE THREAD

**Issue by lmbaeza** [dylni/process_control#7]
| Name            | About                                                                                            | Title             | Label   |
|-----------------|--------------------------------------------------------------------------------------------------|-------------------|---------|
|Propose a change that introduces a new functionality that allows detecting when the memory limit of a process has been exceeded and if it exceeds it, killing the process. | Memory Limit Exceeded Support | request |

It is very useful when you are going to execute processes in which the established memory limit can be exceeded, for example when you execute a process that runs a programming code of some user, such as a Virtual Programming Judge.

```rust
let process = Command::new("echo")
    .arg("hello")
    .stdout(Stdio::piped())
    .spawn()?;

process
    .with_output_timeout(Duration::from_secs(1))
    .memory_limit_exceeded(512MB)
    .terminating()
    .wait()?
    .kill_by_memory_limit_exceeded(|| {
        // Process Memory Limit Exceeded
    })
   .kill_by_timeout(|| {
        // Process timed out
    })?;
```

**Comment by dylni** [comment]
This is a great idea. I've looked into it a bit, and it seems fairly straightforward on Windows, as expected. Unix will be more complicated but should be possible too.

However, I should point out that a programming judge would likely need a separate environment for user programs. There will always be a way to circumvent limits from this crate, potentially allowing programs to damage the system. This crate is mostly intended for trusted programs that might unintentionally exceed their limits, such as code run by test harnesses. Does that still work for your use case?

**Comment by lmbaeza** [comment]
Yes, it does still work for my use case

I have mentioned the example of the virtual judge to get the idea across, but I am also running and monitoring a reliable process, which in the worst case can exceed the defined memory limits.

I think that feature can be useful in many other use cases.

**Comment by dylni** [comment]
> I think that feature can be useful in many other use cases.

I agree. This has been implemented in version 3.2.0 and will be released tomorrow. The API doesn't exactly match the one in this issue, since it resembles the syscalls offered by supported platforms. They cause allocation to fail if the memory usage would be above the memory limit.

Also, MacOS isn't currently supported, since it doesn't offer syscalls for limiting memory usage.

**Comment by dylni** [comment]
Version 3.2.0 has been released with [`Control::memory_limit`](https://docs.rs/process_control/3.2.0/process_control/trait.Control.html#tymethod.memory_limit), so I'm closing this issue. Let me know if you have any issues.
